### ISSUE TITLE
FR:  Switch to gitoxide backend instead of libgit2 for the backend

### ISSUE THREAD

**Issue by nyabinary** [jj-vcs/jj#2316]
**Is your feature request related to a problem? Please describe.**
This is obviously a longer term feature request but it would be extremely nice to have a faster implementation of git implemented in jj instead of the current libgit2 backend, the biggest advantages to switching would be performance and the memory safety of the top of my head.

**Describe the solution you'd like**
Gradual transition to gitoxide as the default backend instead of libgit2

**Describe alternatives you've considered**
Keep using libgit2

**Additional context**
gitoxide is dual licensed with MIT and Apache 2.0, making it compatible with this project.

**Comment by martinvonz** [comment]
I suspect it would be easy to switch to gitoxide for reading and writing commits etc. I actually doubt it would have significant impact on performance. It's probably going to be harder to replace libgit2 by gitoxide for some of the other git integration (push/pull, importing/exporting refs).

**Comment by Nyabinary** [comment]
@martinvonz I mean it claims to be faster so hopefully it is, plus a slow rollout to gitoxide seems like a good idea, replacing libgit2 slowly between versions :3

**Comment by thoughtpolice** [comment]
One advantage of Gitoxide, performance or not, is that pure rust code is vastly easier to build and vendor in stranger circumstances; it may even unblock #1997 for instance, which is pretty badly held up on the need for libopenssl-sys to be compiled (which in turn is a dependency of libgit2-rs)

**Comment by nyabinary** [comment]
What's the status on this being possible?

**Comment by martinvonz** [comment]
It's partially done. Most local operations use Gitoxide, but we haven't started using it for push/pull yet mostly because Gitoxide doesn't support push.

**Comment by yuja** [comment]
push and fetch are ported to subprocessing. https://github.com/jj-vcs/jj/pull/5228
